packages/backend-for-frontend/src/services/toolService.ts

```diff
@@ -0,0 +1,465 @@
+
+function handleValidationResults(
+  errs: {
+    clientAssertionErrors?: Array<ApiError<string>>;
+    keyRetrieveErrors?: Array<ApiError<string>>;
+    clientSignatureErrors?: Array<ApiError<string>>;
+    platformStateErrors?: Array<ApiError<string>>;
+  },
+  clientKind?: authorizationApi.ClientKind,
+  eservice?: bffApi.TokenGenerationValidationEService
+): bffApi.TokenGenerationValidationResult {
+  const clientAssertionErrors = errs.clientAssertionErrors ?? [];
+  const keyRetrieveErrors = errs.keyRetrieveErrors ?? [];
+  const clientSignatureErrors = errs.clientSignatureErrors ?? [];
+  const platformStateErrors = errs.platformStateErrors ?? [];
+
+  return {
+    clientKind,
+    eservice,
+    steps: {
+      clientAssertionValidation: {
+        result: getStepResult([], clientAssertionErrors),
+        failures: apiErrorsToValidationFailures(clientAssertionErrors),
+      },
+      publicKeyRetrieve: {
+        result: getStepResult(clientAssertionErrors, keyRetrieveErrors),
+        failures: apiErrorsToValidationFailures(keyRetrieveErrors),
+      },
+      clientAssertionSignatureVerification: {
+        result: getStepResult(
+          [...clientAssertionErrors, ...keyRetrieveErrors],
+          clientSignatureErrors
+        ),
+        failures: apiErrorsToValidationFailures(clientSignatureErrors),
+      },
+      platformStatesVerification: {
+        result: getStepResult(
+          [
+            ...clientAssertionErrors,
+            ...keyRetrieveErrors,
+            ...clientSignatureErrors,
+          ],
+          platformStateErrors
+        ),
+        failures: apiErrorsToValidationFailures(platformStateErrors),
+      },
+    },
+  };
+}
+
+function assertIsConsumer(
+  requesterId: string,
+  keyWithClient: authorizationApi.KeyWithClient
+) {
+  if (requesterId !== keyWithClient.client.consumerId) {
+    throw organizationNotAllowed(keyWithClient.client.id);
+  }
+}
+
+function getStepResult(
+  prevStepErrors: Array<ApiError<string>>,
+  currentStepErrors: Array<ApiError<string>>
+): bffApi.TokenGenerationValidationStepResult {
+  if (currentStepErrors.length > 0) {
+    return bffApi.TokenGenerationValidationStepResult.Enum.FAILED;
+  } else if (prevStepErrors.length > 0) {
+    return bffApi.TokenGenerationValidationStepResult.Enum.SKIPPED;
+  } else {
+    return bffApi.TokenGenerationValidationStepResult.Enum.PASSED;
+  }
+}
+
+async function retrieveKey(
+  {
+    authorizationClient,
+    purposeProcessClient,
+    agreementProcessClient,
+    catalogProcessClient,
+  }: PagoPAInteropBeClients,
+  jwt: ClientAssertion,
+  ctx: WithLogger<BffAppContext>
+): Promise<
+  SuccessfulValidation<ApiKey | ConsumerKey> | FailedValidation<ErrorCodes>
+> {
+  if (!jwt.payload.purposeId) {
+    return {
+      data: undefined,
+      errors: [purposeIdNotFoundInClientAssertion()],
+    };
+  }
+  const keyWithClient = await authorizationClient.token
+    .getKeyWithClientByKeyId({
```

Comment:
Here there was another doubt we did not clarify: doublecheck with @galales that it's correct to perform this call to the authorization-process client instead of calling the authorization-management as done in Scala (we have a client also for that).